src/Endpoints/class-graphql-metadata.php

```diff
@@ -0,0 +1,269 @@
+
+		$publisher_type = array(
+			'description' => __( 'Parse.ly container type for publisher meta object.', 'wp-parsely' ),
+			'fields'      => array(
+				'logo' => array( 'type' => 'String' ),
+				'name' => array( 'type' => 'String' ),
+				'type' => array( 'type' => 'String' ),
+			),
+		);
+		register_graphql_object_type( self::GRAPHQL_PUBLISHER_TYPE, $publisher_type );
+
+		$meta_type = array(
+			'description' => __(
+				'Metadata fields to be rendered in the front-end. They follow Parse.ly\'s metadata structure. See https://www.parse.ly/help/integration/category/metadata',
+				'wp-parsely'
+			),
+			'fields'      => array(
+				'articleSection'   => array( 'type' => 'String' ),
+				'author'           => array( 'type' => array( 'list_of' => self::GRAPHQL_AUTHOR_TYPE ) ),
+				'context'          => array( 'type' => 'String' ),
+				'creator'          => array( 'type' => 'String' ),
+				'headline'         => array( 'type' => 'String' ),
+				'image'            => array( 'type' => self::GRAPHQL_IMAGE_TYPE ),
+				'keywords'         => array( 'type' => array( 'list_of' => 'String' ) ),
+				'mainEntityOfPage' => array( 'type' => self::GRAPHQL_MAIN_ENTITY_TYPE ),
+				'publisher'        => array( 'type' => self::GRAPHQL_PUBLISHER_TYPE ),
+				'thumbnailUrl'     => array( 'type' => 'String' ),
+				'type'             => array( 'type' => 'String' ),
+				'url'              => array( 'type' => 'String' ),
+			),
+		);
+		/**
+		 * Filters the array for the custom type to represent metadata on GraphQL.
+		 *
+		 * @see https://wpgraphqldocs.gatsbyjs.io/functions/register_graphql_object_type/
+		 * @since 3.2.0
+		 *
+		 * @param array $meta_type Array with the fields of the new type.
+		 */
+		$meta_type = apply_filters( 'wp_parsely_graphql_meta_type', $meta_type );
+		register_graphql_object_type( self::GRAPHQL_META_TYPE, $meta_type );
+
+		$container_type = array(
+			'description' => __( 'Parse.ly Metadata root type.', 'wp-parsely' ),
+			'fields'      => array(
+				'version'  => array(
+					'type'        => 'String',
+					'description' => __( 'Revision of the metadata format.', 'wp-parsely' ),
+				),
+				'meta'     => array(
+					'type'        => 'ParselyMeta',
+					'description' => __( 'Structured and filterable metadata.', 'wp-parsely' ),
+				),
+				'rendered' => array(
+					'type'        => 'String',
+					'description' => __(
+						'HTML string containing the metadata. Intended to be rendered in the front-end as is.',
+						'wp-parsely'
+					),
+				),
+			),
+		);
+		register_graphql_object_type( self::GRAPHQL_CONTAINER_TYPE, $container_type );
+	}
+
+	/**
+	 * Register the custom metadata fields so they can be queried in GraphQL.
+	 *
+	 * @since 3.2.0
+	 *
+	 * @return void
+	 */
+	private function register_fields(): void {
+		$options      = $this->parsely->get_options();
+		$object_types = array_unique( array_merge( $options['track_post_types'], $options['track_page_types'] ) );
```

Comment:
This approach works great, but I have an alternative to propose.

First, tying field registration to this setting introduces the possibility of painful breaking schema changes via the UI. Because this option is controlled via the plugin settings, non-technical users probably have no idea that checking or unchecking a box there has side effects in GraphQL. Unchecking the box to track a custom post type would introduce a breaking schema change and could immediately make all queries to GraphQL return 500.

Second, I think this data can be useful even if the post type is not being tracked by Parse.ly. It benefits SEO and makes rendering consistent. Put another way: Turning off "tracking" for a post type shouldn't change how it looks to Google.

I would make two suggestions:

1. Register this field on an interface instead of on individual post types. The field would be consistently available across all post types that implement the interface (a schema-level decision) and be immune to changes in Parse.ly plugin settings (generally, a UI-level decision).

2. Provide an additional boolean field named `isTracked` that informs whether the option to track this post type is enabled in the UI.

Here are the interfaces that `Post` implements:

<img width="344" alt="Screen Shot 2022-03-09 at 6 58 09 PM" src="https://user-images.githubusercontent.com/739304/157559785-d32eb6f1-1e72-422b-8303-73a4963127e5.png">

While this meta conceptually spans a few of these interfaces, I think `ContentNode` is most appropriate since it is the least common denominator for content that can be described by schema.org data.